owncloudApp/src/main/java/com/owncloud/android/presentation/ui/authentication/LoginActivity.kt

```diff
@@ -455,51 +407,74 @@ class LoginActivity : AppCompatActivity(), SslUntrustedCertDialog.OnSslUntrusted
     }
 
     private fun handleGetAuthorizationCodeResponse(intent: Intent) {
-        val authorizationResponse = AuthorizationResponse.fromIntent(intent)
-        val authorizationException = AuthorizationException.fromIntent(intent)
-        if (authorizationResponse?.authorizationCode != null) {
-            exchangeAuthorizationCodeForTokens(authorizationResponse)
-        } else if (authorizationException != null) {
+        val authorizationCode = intent.data?.getQueryParameter("code")
+
+        if (authorizationCode != null) {
+            Timber.d("Authorization code received [$authorizationCode]. Let's exchange it for access token")
+            exchangeAuthorizationCodeForTokens(authorizationCode)
+        } else {
+            val authorizationError = intent.data?.getQueryParameter("error")
+            val authorizationErrorDescription = intent.data?.getQueryParameter("error_description")
+
+            Timber.e("OAuth request to get authorization code failed. Error: [$authorizationError]. Error description: [$authorizationErrorDescription]")
+            val authorizationException =
+                if (authorizationError == "access_denied") UnauthorizedException() else Throwable()
             updateOAuthStatusIconAndText(authorizationException)
-            Timber.e(authorizationException, "OAuth request to get authorization code failed")
         }
     }
 
     /**
      * OAuth step 2: Exchange the received authorization code for access and refresh tokens
      */
-    private fun exchangeAuthorizationCodeForTokens(authorizationResponse: AuthorizationResponse) {
+    private fun exchangeAuthorizationCodeForTokens(authorizationCode: String) {
```

Comment:
It would be great to try to delegate this sort of responsibilities to the viewmodel... could part of this method be moved to the `LoginViewModel`, keeping here just the observe parts?

Reply:
Yes, I agree. I kept it the same way it was implemented, but I think I should move it. It would be easier to test.